### PR TITLE
Refactor CTC APIs

### PR DESCRIPTION
**Pull request by james77777778**
This PR:
- Ensure dtype inference for `ctc_loss` across all backends
- Add `ctc_loss` to numpy backend
- Add `ctc_decode` (including both `"greedy"` and `"beam_search"` strategies) to numpy backend
- Add `"greedy"` strategy for `ctc_decode` to torch backend
- Reduce the number of import lines by using `keras.src.backend`

Furthermore, we can simply utilize `tf.nn.gelu` instead of a custom implementation since we have already removed the usage of `enable_numpy_behavior`

### CODE REVIEW

**Comment by james77777778** [comment]
UPDATES:
- Update the `ctc_loss` implementation in both jax and numpy

I'm not sure about the reference of the original implementation, but I've found that `optax.ctc_loss_with_forward_probs` is much faster. Therefore, I've made modifications based on it.

|master|this PR|
|-|-|
|~4.7s|~2.0s|

- RTX 4070
- CUDA 12.3

Benchmark script:
```python
import time

import jax

from keras import random
from keras.src.backend.jax.nn import ctc_loss

batch_size = 128
steps = 1024
classes = 256
labels = random.randint(shape=[batch_size, steps], minval=1, maxval=classes)
outputs = random.normal(shape=[batch_size, steps, classes])
length = random.randint(shape=[batch_size], minval=steps // 2, maxval=steps - 1)


fn = jax.jit(ctc_loss)

# Warmup
loss = fn(labels, outputs, length, length)
jax.block_until_ready(loss)

# Benchmark
st = time.time()
for _ in range(100):
    loss = fn(labels, outputs, length, length)
    jax.block_until_ready(loss)
ed = time.time()

print(f"{ed-st:.3f}s")

```

cc @MaanasArora

**Comment by james77777778** [comment]
I've found that the current implementation of `ctc_decode` mismatch with Keras 2:
https://github.com/keras-team/tf-keras/blob/master/tf_keras/backend.py#L7181-L7232

It lacks post-processing of the sparse tensor and doesn't set the blank label as `-1`.

As a result, I've refactored `ctc_decode` in all backends.
Additionally, the jax version of `ctc_decode` is jittable and it should be much faster than before.

I can implement the tf version using low-level ops to support `tf.function(jit_compile=True)` in the next PR.

cc @MaanasArora

**Comment by MaanasArora** [comment]
These changes look great -- thank you!

File: keras/src/backend/tensorflow/nn.py [hunk 1]
```diff
@@ -75,30 +73,7 @@ def selu(x):
 
 def gelu(x, approximate=True):
     x = convert_to_tensor(x)
-    # we need to explicitly implement gelu because bfloat16 will trigger
-    # DTypePromotionError when using enable_numpy_behavior()
```
Comment (fchollet) [inline keras/src/backend/tensorflow/nn.py:79]: Is this no longer a concern?

**Review by fchollet — APPROVED** [review]
LGTM -- thank you for the fixes!
